lib/plugins/stackoverflow/post.dart

```diff
@@ -32,11 +35,86 @@ class StackOverflowPostOwner with _$StackOverflowPostOwner {
       _$StackOverflowPostOwnerFromJson(json);
 }
 
-String postToDiscordMarkdown(StackOverflowPost post) {
-  return '# ${post.title}\n'
+Future<String> postToDiscordMarkdown(
+  StackOverflowPost post,
+  Environment env,
+) async {
+  String header = '# ${post.title}\n'
       '[${post.owner.displayName}](<${post.owner.link}>) - '
       '${timeago.format(DateTime.fromMillisecondsSinceEpoch(post.creationDate * 1000))} | '
       '${post.score} votes | '
-      '[Source](<${post.link}>)\n\n'
-      '${post.bodyMarkdown}';
+      '[Source](<${post.link}>)\n\n';
+
+  String bodyMarkdown = env.pastebinApiKey == null
+      ? post.bodyMarkdown
+      : await shortenWithPastebin(
+          post.bodyMarkdown,
+          PastebinClient(env.pastebinApiKey!),
+          kMaxMessageLength - header.length,
+        );
+
+  String result = '$header$bodyMarkdown';
+
+  if (result.length > kMaxMessageLength) {
+    return '${result.substring(0, kMaxMessageLength - 3)}...';
+  }
+
+  return result;
+}
+
+// Matches code blocks fenced with ```code``` in group 1 and code blocks
+// indented by four spaces in group 2 (SO seems to use the latter quite often).
+final codeblockPattern = RegExp(
+  r'(?:```(?:dart|)\n((?:.*\n)+?)```)|(?:( {4}.+\n(?:(?: {4}.+\n|\n)*) {4}.+)\n)',
+  caseSensitive: false,
+);
+
+Future<String> shortenWithPastebin(
+  String content,
+  PastebinClient client,
+  int maxLength,
+) async {
+  content = content.replaceAll('\r\n', '\n');
+
+  final allMatches = codeblockPattern.allMatches(content);
+  // (start, length, content)
+  final blocks = List.of(allMatches.map((m) => (
+        m.start,
+        m.end - m.start,
+        m.group(1) ?? m.group(2)!,
+      )))
+    // Ignore code blocks less than 200 characters long.
+    ..removeWhere((element) => element.$2 < 200);
+
+  for (final block in blocks) {
+    // Cutting codeblocks that will be cut anyway has no effect
+    if (block.$1 > maxLength) break;
+    // We've managed to cut out enough blocks to make the content fit
+    if (content.length <= maxLength) break;
+
+    final paste = await client.upload(block.$3, language: 'dart');
+
+    final oldLength = content.length;
+
+    content = content.replaceRange(block.$1, block.$1 + block.$2, '''
+[[Long content cut. See full file here]](<$paste>).
+```dart
+${block.$3.split('\n').take(3).join('\n')}
```

Comment:
this has no limit on character length from what I can see,
meaning it could be broken by just adding a _really really_ long line and then no further lines, 
is that right? we should probably account for that.

Reply:
It wouldn't _break_ per se (everything works even if there is a really really long line) and I think it's just more aesthetically pleasing to have three lines instead of X characters.

I think it's quite unlikely this actually happens, so I think this is fine.

Reply:
yes, x lines is more aesthetically pleasing I agree, however we should still hard limit the characters imo, 
because otherwise we get a sort of desync between us limiting code blocks to 200 chars but also us taking 3 lines of the codeblocks.

Reply:
Added a second limit to 200 chars.